Add feed variants with shuffle post-ranking

benhalpern opened the pull request:
<!--
     For Work In Progress Pull Requests, please use the Draft PR feature,
     see https://github.blog/2019-02-14-introducing-draft-pull-requests/ for further details.

     For a timely review/response, please avoid force-pushing additional
     commits if your PR already received reviews or comments.

     Before submitting a Pull Request, please ensure you've done the following:
     - Read the DEV Contributing Guide and the Code of Conduct
     - Provided tests for your changes
     - Used descriptive commit messages
     - Updated any relevant documentation and added any necessary screenshots
-->

## What type of PR is this? (check all applicable)

- [ ] Refactor
- [x] Feature
- [ ] Bug Fix
- [ ] Optimization
- [ ] Documentation Update

## Description
This adds some new variants which take the approach of not randomizing at all _during_ the ranking progress, but entirely shuffling the top half of the results after ranking... This will mean way more variation from the top of the list so new top articles on each load.

But all the randomness is _after_ these are deemed to be top results for the user.

benhalpern commented on app/services/articles/feed.rb:
```diff
@@ -1,6 +1,6 @@
 module Articles
   class Feed
-    RANDOM_OFFSET_VALUES = [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 1, 1, 1, 2, 2, 2, 3, 3, 4, 5, 6, 7, 8, 9, 10, 11].freeze
+    RANDOM_OFFSET_VALUES = [0, 0, 0, 0, 0, 0, 0, 0, 1, 1, 1, 1, 2, 2, 2, 3, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13].freeze
```
Unrelated but I felt like this could be a little less top-heavy

benhalpern commented on app/services/articles/feed.rb:
```diff
@@ -176,7 +205,7 @@ def score_comments(article)
     def globally_hot_articles(user_signed_in)
       hot_stories = published_articles_by_tag.
-        where("score > ? OR featured = ?", 8, true).
+        where("score > ? OR featured = ?", 7, true).
```
This is just switching this magic number down by one, one more tweak not 100% related to the PR, but I think this will be good.

joshpuetz commented on config/field_test.yml:
```diff
@@ -9,15 +9,21 @@ experiments:
+      - more_tag_weight_randomized_at_end_experiment
+      - more_experience_level_weight_randomized_at_end_experiment
+      - more_comments_randomized_at_end_experiment
     weights:
```
Getting busy in here! Any old experiments that could be pruned?

joshpuetz commented on spec/services/articles/feed_spec.rb:
```diff
@@ -197,6 +197,36 @@
+
+    it "includes stories" do
+      expect(stories).to include(old_story)
+      expect(stories).to include(new_story)
```
How valuable are these tests? Perhaps we should exempt these experiment methods from CodeClimate's test coverage percentage check

mstruve replied:
Could extract this into a shared example that loops through each experiment and ensures it does not error and returns an article.

benhalpern replied:
I think that makes sense. I feel like there is some regression test benefit to ensure these keep responding with articles, but I'll refactor into a single simpler spec.

mstruve commented on app/services/articles/feed.rb:
```diff
@@ -115,11 +115,40 @@ def mix_of_everything_experiment
+
+    def more_experience_level_weight_randomized_at_end_experiment
+      @randomness = 0
+      results = more_experience_level_weight_experiment.shuffle
```
Should this `shuffle` be here and in the method below? Given the PR description I was under the impression we just wanted to shuffle the first half.

benhalpern replied:
Oh yeah. That's a mistake. Nice catch.

benhalpern commented:
Alright @joshpuetz @mstruve I refactored the tests and fixed the erroneous `.shuffle` 

I'm going to leave the removal of poorer-performing tests to another PR in order to have fewer moving pieces given that we haven't removed tests yet and I think it makes sense to isolate that behavior in another PR.
